### PR TITLE
Personal/elliswilson/numerous copytool fixes

### PR DESCRIPTION
**Pull request by ellis-wilson-ms**

### CODE REVIEW

**Review by cyrus-ramavarapu-ms — APPROVED** [review]

**Review by tong-xiao-ms — CHANGES_REQUESTED** [review]

File: cmd/lhsmd/agent/action_stats.go [hunk 1]
```diff
@@ -79,6 +80,11 @@ func (as *ActionStats) StartAction(a *Action) {
 func (as *ActionStats) CompleteAction(a *Action, rc int) {
 	s := as.GetIndex(int(a.aih.ArchiveID()))
 	s.queueLength.Dec(1)
+  // Cancel's don't call CompleteAction for themselves.  The action being
+  // canceled must decrement on the cancel action's behalf.
+  if (rc == int(syscall.ECANCELED)) {
```
[the remark this answers is not in the document]
Comment (cyrus-ramavarapu-ms) [reply cmd/lhsmd/agent/action_stats.go:85]: I like this too and it feels doable in StartAction with a similar change to what you did in agent_action.go (check for llapi.HsmActionCancel). Imports will need an update as well.

**Review by cyrus-ramavarapu-ms — CHANGES_REQUESTED** [review]

File: cmd/lhsm-plugin-az/mover.go [hunk 2]
```diff
@@ -286,33 +282,39 @@ func (m *Mover) Archive(ctx context.Context, action dmplugin.Action) error {
...
-	start := time.Now()
-	if action.UUID() == "" {
-		return errors.Errorf("Missing file_id on action %d", action.ID())
+	// translate the fid into an actual path first
+	fidStr := strings.TrimPrefix(action.PrimaryPath(), ".lustre/fid/")
+	fid, err := lustre.ParseFid(fidStr)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse fid")
 	}
-	_, srcObj, err := m.fileIDtoContainerPath(action.UUID())
```
[the remark this answers is not in the document]
Comment (ellis-wilson-ms) [reply cmd/lhsm-plugin-az/mover.go:311]: Good catch.  Removed.

File: cmd/lhsmd/agent/action_stats.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ellis-wilson-ms) [reply cmd/lhsmd/agent/action_stats.go:85]: @cyrus-ramavarapu-ms and @tong-xiao-ms : the reason why I'd like to preserve a cancel action showing up in the queue length is for debugability.  At present all we really get out of the copytool during normal non-error function is:

Oct  5 17:15:38 cb3b6dd2-b409-4c9f-9461-49f67887d1e9-agtpri-a0-vm lhsmd[10760]: 2023/10/05 17:15:38 lhsm-plugin-az2 2023/10/05 17:15:38 total 562,702 (1 min/5 min/15 min/inst): 102/99/97/75 msg/sec
Oct  5 17:15:38 cb3b6dd2-b409-4c9f-9461-49f67887d1e9-agtpri-a0-vm lhsmd[10760]: 2023/10/05 17:15:38 archive:1 total:562,701 queue:1 102/99/97 min:88.310232ms max:744.263542ms mean:366.051947ms median:360.63731ms 75%:500.732598ms 95%:588.236959ms 99%:705.947727ms 99.9%:743.835803ms

If a cancel comes in and somehow gets stuck canceling an archive or restore, then at least I will see that there are two commands in the queue.  In the absence of that lone breadcrumb cancels become totally silent actions.

Accordingly, I'd like to punt on this change request.  Please let me know if either of you object.

File: cmd/lhsmd/agent/agent_action.go [hunk 3]
```diff
@@ -193,26 +151,12 @@ func (action *Action) Update(status *pb.ActionStatus) (bool, error) {
...
-			fileid.URL.UpdateByFid(action.agent.Root(), action.aih.Fid(), []byte(status.Url))
-		}
 		action.agent.stats.CompleteAction(action, int(status.Error))
 		err := action.aih.End(status.Offset, status.Length, 0, int(status.Error))
 		if err != nil {
 			audit.Logf("id:%d completion failed: %v", status.Id, err)
 			return true, err // Completed, but Failed. Internal HSM state is not updated
 		}
-		if action.aih.Action() == llapi.HsmActionArchive && action.agent.config.Snapshots.Enabled && status.Uuid != "" {
-			createSnapshot(action.agent.Root(), action.aih.ArchiveID(), action.aih.Fid(), []byte(status.Uuid))
```
[the remark this answers is not in the document]
Comment (ellis-wilson-ms) [reply cmd/lhsmd/agent/agent_action.go:214]: Good point.  None of the rest of the code uses snapshots or any functions herein anymore, so I've completely blown away the file.

File: cmd/lhsmd/agent_e2e_test.go [hunk 4]
```diff
@@ -43,8 +42,6 @@ var (
 func init() {
 	flag.BoolVar(&enableLeakTest, "leak", false, "enable leak check")
 	flag.Parse()
-	// swap in the dummy implementation
-	fileid.EnableTestMode()
```
[the remark this answers is not in the document]
Comment (ellis-wilson-ms) [reply cmd/lhsmd/agent_e2e_test.go:47]: No.  I removed all support for File Attributes, so there's no longer a need for the test mode.

**Review by tong-xiao-ms — APPROVED** [review]

**Review by cyrus-ramavarapu-ms — APPROVED** [review]

**Pull request merged**